_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,499 @@
+  end
+end
+```
+
+Now that our authentication application is prepped to return encrypted UUIDs (when provided a valid email and password), we
+can create our Social Networking application here's some boilerplate:
+
+##### Our Social Network's user class
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+
+  def self.create_from_authentication_service(email, password)
+    begin
+      connection =  Faraday.new(url: "https://id.wework.io") do |faraday|
+                      faraday.adapter Faraday.default_adapter    
+                      faraday.headers['Content-Type'] = 'application/json'
+                    end
+
+      response = connection.post("/api/v1/login/", {
+                    email: email,
+                    password: password
+                  }.to_json)
+
+      parsed_body = JSON.parse(response.body)
+
+      User.create({
+        uuid:  parsed_body["user"]["uuid"],
+        email: parsed_body["user"]["email"],
+        name:  parsed_body["user"]["name"],
+      })
+    rescue 
+      false
+    end
+  end
+end
+```
+
+##### Our Social Network login controller
+
+```ruby
+class LoginController < ApplicationController
+  
+  def new # Get /login
+  end
+
+  def create  # Post/login
+
+    user = User.create_from_authentication_service(params[:email], params[:password])
+
+    if user
+      cookies.signed[:uuid] = user.uuid
+      redirect_to "some home", notice: "Great Success!"
+    else
+      redirect new_login_path, error: "Something went wrong!"
+    end
+  
+  end
+
+end
+
+```
+
+##### Our Social Network application controller
+
+```ruby
+class ApplicationController < ActionController::Base
+  
+  def current_user
+    @current_user ||= User.find_by(uuid: cookies.signed[:uuid])
+  end
+  helper_method :current_user
+
+end
+
+```
+
+##### Our new Social Network login view
+```html
+<html>
+  <head>
+    <title>Network</title>
+  </head>
+  <body>
+    <form method="POST" action="<%= create_login_path %>">
+      <input type="text" name="email" placeholder="Email" />
+      <input type="password" name="password" placeholder="password" />
+      <input type="submit" value="submit" />
+    </form>
+  </body>
+</html>
+```
+---
+
+Here's the important part:
+
+##### Our new Social Network's Application Controller
+
+```ruby
+class ApplicationController < ActionController::Base
+
+  protected
+
+  def decrypt(encrypted_uuid)
+    begin
+      decipher         = OpenSSL::Cipher.new 'aes-256-cbc'
+      decipher.decrypt
+      decipher.padding = 1
+      decipher.key     = ENV['SOME_SECRET_KEY']
+      decipher.iv      = ENV['SOME_SECRET_IV']
+      decrypted        = cipher.update(Base64.urlsafe_decode64(encrypted_uuid)) + decipher.final
+      decrypted
+    rescue
+      head 403 and return
+    end
+  end
+
+  def current_user
+    if params[:encrypted_user_uuid]
+      @current_user ||= User.find_by_uuid( decrypt_uuid(params[:encrypted_user_uuid]) )
```

Comment:
the revocation is handled at the app level so that all keys would become invalid not just the single users, but would essentially function the same way. Having a single uuid and a single encrypted uuid, would take an awful long time to break it. I'll make some changes to make it more robust..